code/bit-array/src/test/java/ru/unn/agile/bitarray/test/BitArrayTest.java

```diff
@@ -0,0 +1,313 @@
+        assertArrayEquals(reference, actual);
+    }
+
+    @Test
+    public void canUnsetZeroBit() {
+        // arrange
+        BitArray btr = new BitArray(8);
+
+        // act
+        btr.setBit(0);
+        btr.unsetBit(0);
+
+        // assert
+        byte[] actual = btr.getRawArray();
+        byte[] reference = {0};
+        assertArrayEquals(reference, actual);
+    }
+
+    @Test(expected = ArrayIndexOutOfBoundsException.class)
+    public void canUnsetNegativeBitInBitArray() {
+        // arrange
+        BitArray btr = new BitArray(41);
+
+        // act & assert
+        btr.unsetBit(-31);
+    }
+
+    @Test
+    public void isSetBitInBitArray() {
+        // arrange
+        BitArray btr = new BitArray(41);
+
+        // act
+        btr.setBit(29);
+
+        // assert
+        assertTrue(btr.isBit(29));
+    }
+
+    @Test
+    public void isSetBitEqualZeroInBitArray() {
+        // arrange
+        BitArray btr = new BitArray(41);
+
+        // act
+        btr.setBit(0);
+
+        // assert
+        assertTrue(btr.isBit(0));
+    }
+
+    @Test
+    public void isUnsetBitInBitArray() {
+        // arrange
+        BitArray btr = new BitArray(41);
+
+        // act
+        btr.setBit(29);
+
+        // assert
+        assertFalse(btr.isBit(28));
+    }
+
+    @Test(expected = ArrayIndexOutOfBoundsException.class)
+    public void isSetNegativeBitInBitArray() {
+        // arrange
+        BitArray btr = new BitArray(41);
+
+        // act & assert
+        btr.unsetBit(-31);
+    }
+
+    @Test
+    public void isEqualBitArrays() {
+        // arrange
+        BitArray bitArray1 = new BitArray(41);
+        BitArray bitArray2 = new BitArray(41);
+
+        bitArray1.setBit(7);   bitArray2.setBit(7);
+        bitArray1.setBit(11);  bitArray2.setBit(11);
+        bitArray1.setBit(19);  bitArray2.setBit(19);
+        bitArray1.setBit(26);  bitArray2.setBit(26);
+        bitArray1.setBit(34);  bitArray2.setBit(34);
+        bitArray1.setBit(40);  bitArray2.setBit(40);
+
+        // act && assert
+        assertEquals(bitArray1, bitArray2);
+    }
+
+    @Test
+    public void isNotEqualBitArrays() {
+        // arrange
+        BitArray btr1 = new BitArray(41);
+        BitArray btr2 = new BitArray(41);
+
+        btr1.setBit(7);   btr2.setBit(7);
+        btr1.setBit(11);  btr2.setBit(11);
+        btr1.setBit(19);  btr2.setBit(19);
+        btr1.setBit(26);  btr2.setBit(27);
+        btr1.setBit(34);  btr2.setBit(34);
+        btr1.setBit(40);  btr2.setBit(40);
+
+        // act && assert
+        assertNotEquals(btr1, btr2);
+    }
+
+    @Test
+    public void canCopyBitArray() {
+        // arrange
+        BitArray btr = new BitArray(41);
+
+        btr.setBit(4);
+        btr.setBit(13);
+        btr.setBit(19);
+        btr.setBit(25);
+        btr.setBit(34);
+        btr.setBit(41);
+
+        // act
+        BitArray actualBtr = new BitArray(btr);
+
+        // assert
+        assertEquals(btr, actualBtr);
```

Comment:
Did you tried to change second BitArray after creation and check that those not equal? I have feeling, that `rawArray.getRawArray()` will return not a copy, but a reference.

Reply:
@Lenferd corrected